### PR TITLE
Chore/run 2022

### PR DESCRIPTION
**Pull request by tm-danna-ang**
### What does this PR do?
- Reran PH 2017 and ran PH 2022 (with population now as a feature)
- Updates links for ISO3 codes
- Updated `geoBoundaries` link and selected updated column in the GeoJSON (`simplifiedGeometryGeoJSON`)
- Updated `rio-cogeo` and added `pygeos` for new code to work
- Updated nighttime lights to resolve download link for 2023

### Where should reviewers start?
- 2022 run: `notebooks/2024-04-30-single-country-rollouts/ph/`
- 2017 rerun: `notebooks/2023-02-21-single-country-rollouts/ph/`

### CODE REVIEW

File: povertymapping/hdx.py
Comment (joshuacortez) [inline povertymapping/hdx.py:215]: Might be be clearer if the `hdx` utils were renamed to `hrsl` instead, since `hdx` is a huge repository containing many other types of datasets apart from hrsl population. Here we're only considering HRSL population

File: povertymapping/hdx.py
Comment (joshuacortez) [inline povertymapping/hdx.py:236]: not urgent: could we also include `mean` for population density? I think we can pass a list to `func` and `output` 

We can wait a bit to implement this before we merge the PR

Comment (joshuacortez) [reply povertymapping/hdx.py:236]: since this function takes in arbitrary AOIs (not necessarily quadkeys), might be "best practice" to include density by default also

File: povertymapping/feature_engineering.py
Comment (joshuacortez) [inline povertymapping/feature_engineering.py:98]: might have to update this later for adding population density

nit: might be easier to read with just an if statement, no need for the else :)

Comment (joshuacortez) [reply povertymapping/feature_engineering.py:98]: a cleaner implementation is to simply get all features (population not being optional anymore) but agree with this implementation for backwards compatibility!

File: povertymapping/hdx.py
Comment (joshuacortez) [inline povertymapping/hdx.py:218]: might be better to use `None` as the default argument for `extra_args` instead of a dictionary. Using mutable Python objects as a default value for Python functions (dicts and lists are mutable) can lead to unexpected bugs later down the line lol


More detail described [in this blog post](https://docs.python-guide.org/writing/gotchas/), and [here](https://www.geeksforgeeks.org/use-mutable-default-value-as-an-argument-in-python/). Recommended approach is to use `None` in the default value an then define the mutable object (in this case dictionary) within the function

File: notebooks/2024-04-30-single-country-rollouts/ph/3_ph_rollout_model.ipynb
Comment (joshuacortez) [inline notebooks/2024-04-30-single-country-rollouts/ph/3_ph_rollout_model.ipynb:1]: nit: in the last cell, might be better to make the ookla year a variable corresponding to the actual ookla year instead of 2019

or you can delete the cell entirely lol

File: povertymapping/hdx.py
Comment (tm-danna-ang) [reply povertymapping/hdx.py:215]: Makes sense! To clarify which `hdx` utils are you referring to that can be renamed?
- the `hdx.py` filename
- the `generate_hdx_features()` function
- all function and variable names in `hdx.py` that have `hdx` but actually mean `hrsl`

File: povertymapping/hdx.py
Comment (tm-danna-ang) [reply povertymapping/hdx.py:236]: Alright! To clarify, taking the `mean` in this case is already the population density, right?

File: povertymapping/feature_engineering.py
Comment (tm-danna-ang) [reply povertymapping/feature_engineering.py:98]: Got this. Adding `POP_VARIABLES` in `hdx.py` and using an `if` statement instead.

File: povertymapping/hdx.py
Comment (tm-danna-ang) [reply povertymapping/hdx.py:218]: Interesting. I thought that all variables in functions are local but that doesn't seem to be the case for mutable default arguments.

I copied this default arg from `compute_raster_stats()` of `rollout_grids.py` so should I edit that as well, right? Noting here other cases with `list` or `dict` as default args:

`dhs.py`
- `DHSDataManager.__init__()`

`rollout_grids.py`
- `get_region_filtered_bingtile_grids()`
- `generate_dhs_cluster_level_data()`
- `generate_dhs_household_level_data()`
- `generate_dhs_cluster_level_data()`

`nightlights.py`
- `generate_nightlights_feature()`

File: notebooks/2024-04-30-single-country-rollouts/ph/3_ph_rollout_model.ipynb
Comment (tm-danna-ang) [reply notebooks/2024-04-30-single-country-rollouts/ph/3_ph_rollout_model.ipynb:1]: Thanks for the catch! I'll use an f-string for the Ookla year.

File: povertymapping/hdx.py
Comment (joshuacortez) [reply povertymapping/hdx.py:215]: I was thinking renaming all `hdx` to `hrsl` since `hdx` isn't really a dataset but more of a general data hosting service

Do you have in mind specific functions/names/variables can be better kept as `hdx`?

File: povertymapping/hdx.py
Comment (joshuacortez) [reply povertymapping/hdx.py:236]: yes! in that case might be clear also to call it `population_density` and `total_population` vs `pop_mean` and `pop_sum` etc haha

File: povertymapping/hdx.py
Comment (joshuacortez) [reply povertymapping/hdx.py:218]: > I thought that all variables in functions are local but that doesn't seem to be the case for mutable default arguments.

Yeah it is still local! There's a slight nuance hehe. If it's defined within the function then it's local to the function. The `extra_args` is considered local, the mutable object was "created" globally and persists globally. 

Very slight nuances in mutating a variable vs assigning a variable sample here from a [reddit thread](https://www.reddit.com/r/learnpython/comments/wntguq/why_are_only_mutable_variables_a_problem_in/)
```
def func(a=[]):
a = a + [3]
return a

func()
> [3]
func()
> [3]

def func(a=[]):
a += [3]
return a

func()
> [3]
func()
> [3, 3]
```

Comment (joshuacortez) [reply povertymapping/hdx.py:218]: Although I also just realized maybe just using `None`  is not the straightforward solution since there could be multiple other arguments haha

Comment (joshuacortez) [reply povertymapping/hdx.py:218]: It ok to use a dictionary as an input parameter as long as the default value is `None`, something like 
```
---
extra_args: dict | None = None
): 
if extra_args = None:
   extra_args = dict(nodata=np.nan)
```

Comment (joshuacortez) [reply povertymapping/hdx.py:218]: hmmm regarding the other functions that also use a dictionary as a default value, I think would be good to also modify! But lmk if modifying will be cumbersome, in that case we can file an issue and address it in another PR instead

File: povertymapping/hdx.py
Comment (tm-danna-ang) [reply povertymapping/hdx.py]: I recommend to retain these as `hdx` since they're the config for calling the HDX Python API:
- `HDX_CONFIG`
- `init_hdx_config()`

All other cases, seems more appropriate to be `hrsl`:
- `hdx.py`
- `download_hdx()`
- `get_unzipped_hdxfile()`
- `get_hdx_file()`

File: povertymapping/hdx.py
Comment (tm-danna-ang) [reply povertymapping/hdx.py]: From our discussions, retained only `population_density` as the new feature.

File: povertymapping/feature_engineering.py
Comment (tm-danna-ang) [reply povertymapping/feature_engineering.py]: Changes in this thread are reverted as previously I used that workaround since the `pop_count` column was both a feature and part of the `input_cols`. With the new `population_density` column, this workaround is not needed anymore.

File: povertymapping/hdx.py
Comment (tm-danna-ang) [reply povertymapping/hdx.py]: Went over them and the fix is simple enough for the 7 functions (`DHSDataManager.__init__()` is a false alarm) so can do it in this PR.

File: povertymapping/hdx.py
Comment (joshuacortez) [reply povertymapping/hdx.py]: oh there are utilities using the HDX python API, oops I missed that. Thanks!

**Review by joshuacortez — APPROVED** [review]
should be good to merge after all the edits!

**Pull request merged**
